### PR TITLE
TaskInstantiation integration test

### PR DESCRIPTION
**Pull request by bjoernschroeder**

### CODE REVIEW

File: alica_engine/src/engine/Assignment.cpp [hunk 1]
```diff
@@ -249,9 +249,9 @@ AssignmentView Assignment::getAgentsWorking(int idx) const
     return AssignmentView(this, idx);
 }
 
-AssignmentView Assignment::getAgentsWorking(int64_t epId) const
+AssignmentView Assignment::getAgentsWorking(int64_t epId, int64_t dynamicId = 0) const
```
Comment (veerajsk) [inline alica_engine/src/engine/Assignment.cpp]: the default argument is usually in the header. I'm not sure that this works

File: alica_tests/src/AssignPayloadSummand.cpp [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        }
+    }
+    std::vector<double> distances;
+    for (auto it = nonVisitedEps.begin(); it != nonVisitedEps.end(); it++) {
+        for (AgentId agentId : ass.getAgentsWorking(movePayloadEpId, *(it))) {
+            int x = worldModel->agentLocations.at(agentId).first - worldModel->payloads[*it - 1].pickX;
+            int y = worldModel->agentLocations.at(agentId).second - worldModel->payloads[*it - 1].pickY;
+            double distance = sqrt(pow(x, 2) + pow(y, 2));
+            if (distance < 1) {
+                distance = 1;
```
Comment (veerajsk) [inline alica_tests/src/AssignPayloadSummand.cpp:59]: why is this required?

File: alica_tests/src/AssignPayloadSummand.cpp [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    for (uint64_t dynId = 1; dynId < 9; dynId++) {
+        auto agentsInEntryPoint = ass.getAgentsWorking(movePayloadEpId, dynId);
+
+        // dont accept assignments with multiple agents in one entry point
+        if (agentsInEntryPoint.size() > 1) {
+            return UtilityInterval(-1, -1);
+        }
+
+        // agents should not be assigned to payloads that have already been moved
+        if (agentsInEntryPoint.size() > 0 && worldModel->payloads.at(dynId - 1).state == alicaTests::PayloadState::DROPPED) {
```
Comment (veerajsk) [inline alica_tests/src/AssignPayloadSummand.cpp]: can use payloadId instead of dynId - 1

File: alica_tests/src/test/test_alica_dynamic_task.cpp [hunk 4]
```diff
@@ -133,6 +134,136 @@ class AlicaDynamicTaskPlanTest : public AlicaTestMultiAgentFixture
...
+                std::make_unique<alica::ConstraintCreator>(), std::make_unique<alica::BehaviourCreator>(), std::make_unique<alica::PlanCreator>());
+
+        for (int i = 0; i < getAgentCount(); ++i) {
+            alica::AlicaContext* ac =
+                    new alica::AlicaContext(alica::AlicaContextParams(getHostName(i), path + "/etc/", getRoleSetName(), getMasterPlanName(), stepEngine()));
+            ASSERT_TRUE(ac->isValid());
+            cbQueues.emplace_back(std::make_unique<ros::CallbackQueue>());
+            spinners.emplace_back(std::make_unique<ros::AsyncSpinner>(4, cbQueues.back().get()));
+            ac->setCommunicator<alicaDummyProxy::AlicaDummyCommunication>();
+            ac->setWorldModel<alicaTests::TaskInstantiationIntegrationWorldModel>();
```
Comment (veerajsk) [inline alica_tests/src/test/test_alica_dynamic_task.cpp:196]: can create a single instance of the world model that is shared by all the agents. Then set the world model for each agent as a pointer to this single object. Will also need to add a mutex to the world model & any agent that wants to change the world model, does so only after locking the mutex. This will remove the cyclic references & simply a lot of this code

File: alica_tests/src/test/Expr/src/NavigateToDrop.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    /*PROTECTED REGION END*/
+}
+void NavigateToDrop::run(void* msg)
+{
+    /*PROTECTED REGION ID(run4459885370764933844) ENABLED START*/
+    // Add additional options here
+    if (isSuccess()) {
+        return;
+    }
+    _worldModel->agentLocations[_agentId].first = _worldModel->payloads[*(_worldModel->currentPayloadId)].dropX;
```
Comment (veerajsk) [inline alica_tests/src/test/Expr/src/NavigateToDrop.cpp]: shouldn't you be updating all the world models here? Although, with the change of using a single shared model suggested earlier, you won't need to do so.

File: alica_tests/src/test/Expr/include/AssignPayload.h [hunk 6]
```diff
@@ -0,0 +1,32 @@
+#pragma once
```
Comment (veerajsk) [inline alica_tests/src/test/Expr/include/AssignPayload.h]: can get rid of assign/unassign payload behaviours & move their functionality to the the onInit & before setting success of the NavigateToPick & Drop behaviours respectively

File: alica_tests/src/test/Expr/include/NavigateToPick.h [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#pragma once
```
Comment (veerajsk) [inline alica_tests/src/test/Expr/include/NavigateToPick.h:1]: I guess this branch is before blackboard support, otherwise can use a single navigation behaviour with a parameter specifying the target

File: alica_tests/src/AssignPayloadSummand.cpp (hunk 2 above)
Comment (bjoernschroeder) [reply alica_tests/src/AssignPayloadSummand.cpp:59]: An agent that is already standing on the payloads pick position will have a distance of 0, leading to a division by 0 error in the next step. Also, agents with a low distance to the pick spot (for example, 0.001) will increase the utility by a huge amount and values might get close to infinity.

File: alica_tests/src/test/Expr/include/NavigateToPick.h (hunk 7 above)
Comment (bjoernschroeder) [reply alica_tests/src/test/Expr/include/NavigateToPick.h:1]: It already has an older version of the blackboard. However in setParameters, there is no way to access the worldModel, as the PlanAttachment object, on which the setParameters method is called, does not store any ptr to the worldModel / engine and takes neither as an input to setParameters.

File: alica_engine/include/engine/Assignment.h [hunk 8]
```diff
@@ -111,7 +111,7 @@ class Assignment
 
     AssignmentView getAgentsWorking(const EntryPoint* ep) const;
     AssignmentView getAgentsWorking(int idx) const;
-    AssignmentView getAgentsWorking(int64_t epId) const;
+    AssignmentView getAgentsWorking(int64_t epId, int64_t dynamicId = 0) const;
```
Comment (veerajsk) [inline alica_engine/include/engine/Assignment.h:114]: default argument declaration should be in IAssignment
